components/supervisor/pkg/supervisor/top.go

```diff
@@ -17,10 +17,57 @@ import (
 	"google.golang.org/grpc/credentials/insecure"
 
 	linuxproc "github.com/c9s/goprocinfo/linux"
+	"github.com/gitpod-io/gitpod/common-go/log"
 	"github.com/gitpod-io/gitpod/supervisor/api"
 	daemonapi "github.com/gitpod-io/gitpod/ws-daemon/api"
 )
 
+type TopService struct {
+	data *api.ResourcesStatusResponse
+}
+
+func (t *TopService) GetTopData() *api.ResourcesStatusResponse {
+	var newData *api.ResourcesStatusResponse = t.data
+	return newData
+}
+
+func (t *TopService) updateTopData(ctx context.Context) {
+	data, err := Top(ctx)
+	if err != nil {
+		log.WithField("error", err).Error("Failed to retrieve resource status from upstream")
+	}
+	t.data = data
+}
+
+// Observe starts observing the resource status
+func (t *TopService) Observe(ctx context.Context) {
+	var (
+		delay                       = 1 * time.Second
+		reconnectionDelayGrowFactor float64
+		maxReconnectionDelay        = 30 * time.Second
+	)
+
+	t.updateTopData(ctx)
```

Comment:
I also call this immediately otherwise `Observe` waits 1 second before fetching the metrics

Reply:
you can do it like that to avoid code duplications:
```go
for {
      data, err := Top(ctx)
      if err != nil {
            log.WithField("error", err).Error("Failed to retrieve resource status from upstream")
      } else {
           delay = minReconnectionDelay
	   t.data = data
      }
   select {
	case <-ctx.Done():
		log.Info("Resource Status observer stopped")
		return
	case <-time.After(delay):
		delay = time.Duration(float64(delay) * reconnectionDelayGrowFactor)
		if delay > maxReconnectionDelay {
			delay = maxReconnectionDelay
		}
	}
}

Reply:
Updated 🙏